src/lib/storage/zero_suppression/simd_bp128/simd_bp128_decoder.hpp

```diff
@@ -0,0 +1,132 @@
+
+#include "types.hpp"
+
+namespace opossum {
+
+class SimdBp128Vector;
+
+/**
+ * @brief Implements point-access into a SIMD-BP128 compressed vector
+ *
+ * The decoder caches the last decoded block and meta-block.
+ * Performs best if the values are accessed in a cache-coherent manner, i.e.,
+ * sequentially without too many jumps backwards.
+ */
+class SimdBp128Decoder : public BaseZeroSuppressionDecoder {
+ public:
+  using Packing = SimdBp128Packing;
+
+ public:
+  explicit SimdBp128Decoder(const SimdBp128Vector& vector);
+  SimdBp128Decoder(const SimdBp128Decoder& other);
+
+  SimdBp128Decoder(SimdBp128Decoder&& other) = default;
+  ~SimdBp128Decoder() = default;
+
+  uint32_t get(size_t i) final {
+    if (_is_index_within_cached_block(i)) {
+      return _get_within_cached_block(i);
+    }
+
+    if (_is_index_within_cached_meta_block(i)) {
+      return _get_within_cached_meta_block(i);
+    }
+
+    if (_is_index_after_cached_meta_block(i)) {
+      const auto relative_index = _index_within_cached_meta_block(i);
+      const auto relative_meta_block_index = relative_index / Packing::meta_block_size;
+
+      _read_meta_info_from_offset(relative_meta_block_index);
+      return _get_within_cached_meta_block(i);
+    }
+
+    _reset_cached_meta_block();
+    _read_meta_info(_cached_meta_info_offset);
+    const auto meta_block_index = i / Packing::meta_block_size;
+    _read_meta_info_from_offset(meta_block_index);
+    return _get_within_cached_meta_block(i);
+  }
+
+  size_t size() const final { return _size; }
+
+ private:
+  bool _is_index_within_cached_block(size_t index) {
+    const auto begin = _cached_block_first_index;
+    const auto end = _cached_block_first_index + Packing::block_size;
+    return begin <= index && index < end;
+  }
+
+  size_t _index_within_cached_block(size_t index) { return index - _cached_block_first_index; }
+
+  bool _is_index_within_cached_meta_block(size_t index) {
+    const auto begin = _cached_meta_block_first_index;
+    const auto end = _cached_meta_block_first_index + Packing::meta_block_size;
+    return begin <= index && index < end;
+  }
+
+  size_t _index_within_cached_meta_block(size_t index) { return index - _cached_meta_block_first_index; }
+
+  bool _is_index_after_cached_meta_block(size_t index) {
+    return (_cached_meta_block_first_index + Packing::meta_block_size) <= index;
+  }
+
+  uint32_t _get_within_cached_block(size_t index) { return (*_cached_block)[_index_within_cached_block(index)]; }
+
+  uint32_t _get_within_cached_meta_block(size_t index) {
+    const auto block_index = _index_within_cached_meta_block(index) / Packing::block_size;
+    _unpack_block(block_index);
+
+    return (*_cached_block)[_index_within_cached_block(index)];
+  }
+
+  /**
+   * Starting from the cached meta info offset,
+   * jumps to the meta block with the relative
+   * index `meta_block_index`
+   */
+  void _read_meta_info_from_offset(size_t meta_block_index) {
```

Comment:
The meta blocks all have the same size - why would I want to start with the cached one?

Reply:
They don’t. They contain the same amount of elements, but their physical / encoded size depends on the chosen bit size, which changes from block to block.